### ISSUE TITLE
'Load' use-cases step (3)

### ISSUE THREAD

**Issue by dladd** [cellml/libcellml#111]
This issue covers the Load step as outlined in the [use-cases doc](http://libcellml.readthedocs.io/en/latest/usecases.html). This currently includes:
- [x] Updates to the object model
- [x] Load each of the models from 1 and 2 (new models can be added for this part if required).
  - [x] a model with imported components
    - [x] a single component
    - [x] a component with a hierarchy
    - [x] a component from a non-existent URL

The major task for this step is adding `deserialise(std::string)` operations to parse out CellML 1.2 objects. ~~using regular expressions.~~

**Comment by dladd** [comment]
In terms of testing, we could (option A) construct another parallel set of deserialisation test cases- for example:

``` c++
TEST(Component, serialiseValidName) {
    const std::string name = "valid_name";
    const std::string e = "<component name=\"" + name + "\"/>";
    libcellml::Component c;
    c.setName(name);
    std::string a = c.serialise(libcellml::Formats::XML);
    EXPECT_EQ(e, a);
}

TEST(Component, deserialiseValidName) {
    const std::string e = "valid_name";
    const std::string input = "<component name=\"" + e + "\"/>";
    libcellml::Component c = libcellml::Component::deserialise(input);
    std::string a = c.getName;
    EXPECT_EQ(e, a);
}
```

Another option (B) is to simply add a deserialisation/reserialisation step on to the existing test cases:

``` c++
TEST(Component, validName) {
    const std::string name = "valid_name";
    const std::string e = "<component name=\"" + name + "\"/>";
    libcellml::Component c;
    c.setName(name);
    std::string a = c.serialise(libcellml::Formats::XML);
    EXPECT_EQ(e, a);
    libcellml::Component deserialisedC = libcellml::Component::deserialise(e);
    std::string a = deserialisedC.serialise(libcellml::Formats::XML);
    EXPECT_EQ(e, a);
}
```

I think it makes sense to do separate test cases (A) for some of the initial tests while we get this up and running. However, for more complex cases, de/reserialisation (B) will allow us to quickly work through our very thorough testing base with minimal reduplication of effort and additional code to maintain.

**Comment by jonc125** [comment]
Regarding the approach to parsing, is using regular expressions really the way we want to go? There are so many good XML parsing libraries out there, which have been thoroughly tested...

**Comment by agarny** [comment]
> Regarding the approach to parsing, is using regular expressions really the way we want to go? There are so many good XML parsing libraries out there, which have been thoroughly tested...

Regular expressions can be very handy, but... here we would have to use quite a few of them and/or complicated ones, which means that it would slow things quite a bit.

So, personally, I would indeed go with a well-tested XML parsing library.

**Comment by dladd** [comment]
Thanks, @agarny and @jonc125. Re: regex, this was a misunderstanding on my part. From #32, it looks like the consensus was to go with libxml2 from this step on.

**Comment by dladd** [comment]
One thing that might be worth considering is whether to separate out the libxml2 side of this from the rest of the API. The c/c++ differences and manipulations of `xmlChar`s from libxml2 in particular could otherwise make things a bit ugly. I'm in favour of wrapping the libxml2 calls in buffer classes, which I'm currently calling `XmlDoc` and `XmlNode` to correspond with operations on `xmlDoc` and `xmlNode` in libxml2. 

I'm working on a more complete proof-of-concept but for a quick illustrative example:

In model.cpp

``` c++
void Model::doDeserialisation(const XmlNodePtr &node)
{
    // Check that the current XML node has the correct element type
    if (node->hasElement("model")) {
        // Check if this model has a name attribute
        if (node->hasAttribute("name")) {
            // Set the name for this object
            this->setName(node->getAttribute("name"));
        }
        ...
```

and in xmlnode.cpp

``` c++
bool XmlNode::hasElement(const char *elementName)
{
    bool found = false;
    if (!xmlStrcmp(mPimpl->mXmlNodePtr->name, (const xmlChar *) elementName)) found = true;
    return found;
}

bool XmlNode::hasAttribute(const char *attributeName)
{
    bool found = false;
    xmlAttrPtr attribute = xmlHasProp(mPimpl->mXmlNodePtr, (const xmlChar *) attributeName);
    if (attribute) found = true;
    return found;
}

std::string XmlNode::getAttribute(const char *attributeName)
{
    xmlChar *attributeValue = xmlGetProp(mPimpl->mXmlNodePtr, (const xmlChar *) attributeName);
    std::string attributeValueString(reinterpret_cast<char*>(attributeValue));
    return attributeValueString;
}
```

If that's sounding reasonable, I'll crack on with fleshing this out.

**Comment by jonc125** [comment]
Been in meetings all day today so haven't had a chance to look at this properly, but I'd certainly approve of using wrapper classes in principle. You might want to check whether a straight cast is sufficient for strings or whether you need to transcode - certainly for Xerces the XML strings are unicode so you can't put them directly in a char\* or std::string.

**Comment by nickerso** [comment]
Wrapping seems like the way to go to me, as long as those Xml\* classes are neatly hidden internally and never exposed via the libCellML API.

BTW, with libxml2 casting strings is standard practice in all the code I've seen (including their own tests/examples) and they provide a macro `BAD_CAST`, I think, to handle the `char*` to `xmlChar*` cast.

**Comment by dladd** [comment]
Thanks @jonc125 and @nickerso. Good point re: BAD_CAST- that should make this look more readable. It also looks like we have to explicitly free with `xmlFree()`. So my `getAttribute()` above should now be:

``` c++
std::string XmlNode::getAttribute(const char *attributeName)
{
    char* attributeValue = (char*)xmlGetProp(mPimpl->mXmlNodePtr, BAD_CAST attributeName);
    std::string attributeValueString = std::string(attributeValue);
    xmlFree(attributeValue);
    return attributeValueString;
}
```

**Comment by nickerso** [comment]
yep, you need to use libxml to free all the memory it gives you or you'll end up with segfaults all over the place (as I have learnt and re-learnt several times!).

**Comment by dladd** [comment]
For those interested, you can find a working proof-of-concept of this in my [load branch](https://github.com/dladd/libcellml/tree/load) (example [test](https://github.com/dladd/libcellml/blob/load/tests/model/model.cpp#L48), [implementation](https://github.com/dladd/libcellml/blob/load/src/model.cpp#L250), and [wrapper](https://github.com/dladd/libcellml/blob/load/src/xmlnode.cpp)). Please note this is not intended to be complete and also still needs some tweaking around scoping and naming (among the usual nits :-) ).

If there are no objections, I'll next extrapolate these ideas to how they might work for the rest of the API and come back with an update to the object model for review.

**Comment by nickerso** [comment]
I haven't looked too close at the code, but I'm not sure why the xmldoc and xmlnode headers are in the API folder and that `libcellml` module? Shouldn't these be internal use only classes and not exposed?

Otherwise, this seems like a reasonable start to loading models.

**Comment by nickerso** [comment]
actually, now I think I see why they are in the API. The `Component::deserialise` method is taking a `XmlNodePtr` argument. I would expect `deserialise` to always take a string, as the opposite to `serialise`.  But we probably don't want to be continually parsing XML to/from strings....

Perhaps there needs to be a separate parser of some kind which is responsible for all the deserialisation?

**Comment by dladd** [comment]
Thanks @nickerso- you're absolutely right about needing to change the exposure of the XML routines to internal-only. I was really just trying to check that libxml integration worked as I imagined here. I've now popped the headers into src/.

The `Entity::deserialise()` method was overloaded to take either `(std::string, Format)` from the user or `(XmlNodePtr)` for internal hierarchical deserialisation. But that is probably confusing- I can rename the internal method to `deserialiseXmlNode`.

Re: a separate parser versus `deserialise` methods in each of the existing classes, I think I am in favour of the latter as more consistent with our `serialise` methods but I am (as always) amenable to the community decision :)

**Comment by agarny** [comment]
I have just had a quick look at it all and I like the idea of the wrapper. As for a separate parser versus `deserialise` methods, I also prefer the latter (it's consistent with the `serialise` methods indeed).

Otherwise, I find `BAD_CAST` a strange name for that macro. To me, at least, it doesn't convey the right meaning...

**Comment by dladd** [comment]
Re: BAD_CAST, from the [libxml2 doc](http://xmlsoft.org/html/libxml-xmlstring.html#BAD_CAST):

> Macro to cast a string to an xmlChar \* when one know its safe.
> This is a basic byte in an UTF-8 encoded string. It's unsigned allowing to pinpoint case where char \* are assigned to xmlChar \* (possibly making serialization back impossible). 

So I guess they are calling it "bad" to make it clear that it might not be safe for re-serialisation?

**Comment by dladd** [comment]
I was originally hoping to do deserialisation by allowing a user to pass a string to `deserialise()`, and do all the parsing into an `XmlDoc` and `XmlNode`s under the hood. However, an issue I've run into is that this either requires public access to functions that take an `XmlNode` or a `friend class` work around to get access to protected methods.

Currently, I see three options to address the issue:
- **(A)** Expose our `XmlDoc` and `XmlNode` wrappers to the user. The idea here is that if we expose any of our XML handling, we should expose all of it and allow users to manipulate these objects if they choose. A simplified internal implementation for deserialising names from a model and its components would be (as in the previous proof-of-concept):

``` c++
void Model::doDeserialisation(const XmlNodePtr &node)
{
    if (node->isElementType("model")) {
        if (node->hasAttribute("name")) {
            this->setName(node->getAttribute("name"));
        }
        XmlNodePtr childNode = node->getChild();
        while (childNode) {
            if (childNode->isElementType("component")) {
                ComponentPtr component = std::make_shared<Component>();
                component->deserialiseXmlNode(childNode);
                this->addComponent(component);
            }
            childNode = childNode->getNext();
        }
    } else {
        throw std::invalid_argument("Unexpected XML element type.");
    }
}
```
- **(B)** Changing the internal deserialise methods to take string arguments. This would require us to create a new `XmlDoc` object for each entity in a CellML model as we step through the tree, convert the `XmlDoc` to its root `XmlNode`, perform the expected deserialisation for that entity, and convert any child entities to strings to repeat the process on:

``` c++
void Model::doDeserialisation(const std::string &input, Format format)
{
    if (format == Format::XML) {
        XmlDocPtr doc = std::make_shared<XmlDoc>();
        doc->parse(input);
        const XmlNodePtr node = doc->getRootNode();
        if (node->isElementType("model")) {
            if (node->hasAttribute("name")) {
                this->setName(node->getAttribute("name"));
            }
            XmlNodePtr childNode = node->getChild();
            while (childNode) {
                if (childNode->isElementType("component")) {
                    ComponentPtr component = std::make_shared<Component>();
                    std::string componentString = childNode->convertToString();
                    component->deserialise(componentString, format);
                    this->addComponent(component);
                }
                childNode = childNode->getNext();
            }
        } else {
            throw std::invalid_argument("Unexpected XML element type.");
        }
    } else {
        throw std::out_of_range("Unrecognised format specified.");
    }
}
```
- **(C)** Work around the public/private issue by adding `friend class Base` to each Derived class to be deserialised ([model/component example](https://github.com/dladd/libcellml/blob/load/src/api/libcellml/component.h#L214)). Implementation would be similar to (A). While apparently an easy fix, there does seem to be some controversy around using `friend` when inheritance is involved like this.

I am personally in favour of either (A) or (C).

**Comment by nickerso** [comment]
hmmmm...just to be difficult, I'm going to throw in a **(D)**, which is back to my previous idea of a separate parser object. I'd expect to use it like this:

``` c++
libcellml::Parser parser(libcellml::Format::XML);
libcellml::ModelPtr model = std::make_shared<libcellml::Model>();
try {
    parser->parseModel(model, wholeModelString);
    // or maybe
    model = parser->parseModel(wholeModelString);
    // might be better?
} catch (...) {}

libcellml::ComponentPtr component = make_shared<libcellml::Component>();
try {
    parser->parseComponent(component, stringWithComponentOnly);
} catch (...) {}
```

Although to begin with I would only suggest providing the `parseModel()` method. The `Parser` class would only use the public API on the core libCellML objects to populate or create the model. And perhaps the `Parser` class would be an internal object used only by the `deserialise()` methods? Although perhaps useful to provide this as a user-configurable way to load resources in general, could set a base URL to use in resolving relative imports, etc.

Obviously (D) is my preferred option, but otherwise I would favour (C) but I have no idea on the use of `friend` :) (B) is also fine by me for now.

**Comment by agarny** [comment]
I don't think I want something like **(D)**. At this stage, I would much prefer the whole deserialisation to be done at once through one call to the API.

Regarding **(A)**, **(B)** and **(C)**, I can now see the problem with **(A)** and I don't think we want to expose `XmlDoc` and `XmlNode` to the user, so **(A)** is not an option for me. As for **(C)**, well... I am really not a `friend` advocate. I am sure some C++ gurus would say that there cases where it's fine to use `friend`, but personally I can't help but think that it's an easy 'fix' to a bad OO design. So, **(C)** is not an option for me either. This leaves me with **(B)** and, looking at it, I can't see why it's not your first choice @dladd and @nickerso, or am I missing something?

**Comment by nickerso** [comment]
I think its just the thought of continual parsing/serialising strings to XML to strings that bothers me. No technical objection to it. Indeed, seeing more opinion regarding (C), I'd be more inclined to opt for (B) instead. (D) still number one though! ;-)

**Comment by dladd** [comment]
I'm not a fan of (B) because it requires string->XmlDoc->XmlNode->string conversions for every deserialised entity. That also means lots of BAD_CASTing going on under the hood. It just seems a bit excessive to me when we already have the XML tree fully defined from the start.

But I'm happy to go with the consensus :)

**Comment by agarny** [comment]
Hmm, my bad for focusing too much on **(A)** and **(C)**, and not enough on **(B)**. I now see what you mean and I agree that all those conversions are far from ideal.

Otherwise, I am actually puzzled by **(A)**. I get the feeling that it should be possible to have **(A)** without having to expose `XmlDoc` and `XmlNode` to the user.

**Comment by agarny** [comment]
Outcome of a quick chat with @dladd: we should be able to do the whole de/serialisation from the `Model` class, as opposed to doing a particular bit of the de/serialisation in the `Model`, `Component`, `Variable`, etc. classes. This means that we could then have something similar to **(A)** but without having to expose `XmlDoc` and `XmlNode` since they would only be accessed by the `Model` class.

**Comment by dladd** [comment]
To build on @agarny 's comment above, let's say (E) is:
- **(E)** Perform deserialisation from the model class only. An example implementation corresponding with the others above might be:

``` c++
void Model::deserialise(const std::string &input, Format format)
{
    if (format == Format::XML) {
        XmlDocPtr doc = std::make_shared<XmlDoc>();
        doc->parse(input);
        const XmlNodePtr node = doc->getRootNode();
        if (node->isElementType("model")) {
            if (node->hasAttribute("name")) {
                this->setName(node->getAttribute("name"));
            }
            XmlNodePtr childNode = node->getChild();
            while (childNode) {
                if (childNode->isElementType("component")) {
                    ComponentPtr component = loadComponent(childNode);
                    this->addComponent(component);
                }
                childNode = childNode->getNext();
            }
        } else {
            throw std::invalid_argument("Unexpected XML element type.");
        }
    } else {
        throw std::out_of_range("Unrecognised format specified.");
    }
}

ComponentPtr Model::loadComponent(const XmlNodePtr &node)
{
    ComponentPtr component = std::make_shared<Component>();
    if (node->hasAttribute("name")) {
        component->setName(node->getAttribute("name"));
    }
    return component;
}
```

Although I suppose this becoming quite similar to @nickerso 's suggestion of a separate parser- just assuming that we are only parsing models. **(D)** might give us more flexibility to parse out other objects with the possible trade-off of having a user define a parser.

**Comment by jonc125** [comment]
Having done both kinds of approach (A/D) in previous projects, I'd be very much in favour of D.

**Comment by agarny** [comment]
I have just checked what I did back then in COR and well... it's effectively @nickerso's suggestion, except that instead of passing strings, I pass the equivalent of `XmlNode` objects (i.e. similar to what **(E)** does).

So, I am in favour of **(D)**, but with `XmlNode` objects rather than strings.

**Comment by nickerso** [comment]
Its important to remember that XML is just one of potentially several serialisation formats that we want to eventually support. So I'd really like to not see XML nodes in the (public) API if we can avoid it.

**Comment by agarny** [comment]
Hmm, good and fair point. Just one thing then: looking at your snippet, if we want to deserialise the whole model, we would indeed start with `wholeModelString`, but then how would we get `stringWithComponentOnly`?

**Comment by jonc125** [comment]
You'd have pairs of methods on the parser - parseX(string) would be public, allowing users to parse parts of models, and parseX(XmlNode) would be private, used by the public methods to parse each kind of entity. The parseX(XmlNode) methods would only call the private methods to parse child nodes, so there'd be no repeated conversion to/from string.

Of course, the private methods would be hidden entirely so there's not even a mention of XmlNode in the public API.

**Comment by agarny** [comment]
Thanks @jonc125, it makes sense to me now.

**Comment by dladd** [comment]
I assume we want the parser to only instantiate objects below the inheritance level it is called from. I.e. we would still expect a user to do

``` c++
libcellml::Parser parser(libcellml::Format::XML);
libcellml::ModelPtr model = std::make_shared<libcellml::Model>();
parser.parseModel(model, wholeModelString);
```

rather than

``` c++
libcellml::Parser parser(libcellml::Format::XML);
libcellml::ModelPtr model = parser.parseModel(wholeModelString);
```

Both would have to create new objects for components, variables, etc. from the `wholeModelString`. In the former, a user can take objects/attributes from the string and add/overwrite to those from an existing model object, while the second assumes we're creating a new object. 

I suppose another option is to overload `parseModel()` with both methods, allowing the user to do either based on the signature.

**Comment by agarny** [comment]
I would indeed overload `parseModel()` to allow both.

**Comment by nickerso** [comment]
yep, me too. With the rules that any named entity with a matching name is replaced with the new one from the parser, everything else is appended.

**Comment by hsorby** [comment]
I really think that this requires two methods with different names.  I would think something like:

`
libcellml::ModelPtr model = parser.parse(wholeModelString);
`

and

`
libcellml::ModelPtr model // existing model from somewhere
`
`
parser.updateModel(model, modelString)
`

**Comment by agarny** [comment]
I am fine with different names, but it would then have to be `parseModel()` rather than `parse()` since we will also want to have something similar for components, variables, etc.

**Comment by dladd** [comment]
Re:

> With the rules that any named entity with a matching name is replaced with the new one from the parser, everything else is appended.

I assume the appending extends to all encapsulated entities, rather than a straight replacement to the top-level named entity? 

For example, if we had an existing model with `variable1` in `component1`: 

``` xml
            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
            "<model xmlns=\"http://www.cellml.org/cellml/1.2#\">"
                "<component name=\"component1\">"
                    "<variable name=\"variable1\"/>"
                "</component>"
            "</model>";
```

and we did an `updateModel()` with `variable2` in `component1`:

``` xml
            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
            "<model xmlns=\"http://www.cellml.org/cellml/1.2#\">"
                "<component name=\"component1\">"
                    "<variable name=\"variable2\"/>"
                "</component>"
            "</model>";
```

 the resulting model `component1` would contain both `variable1` and `variable2` rather than just `variable2`?

**Comment by agarny** [comment]
> the resulting model `component1` would contain both `variable1` and `variable2` rather than just `variable2`?

That's indeed what I would personally expect to get.

**Comment by nickerso** [comment]
I would expect `component1` to only contain `variable2`, the original `component1` being replaced by the new `component1`.

But I do think for now that we should only be implementing the straight out `parse()` (`parseModel()`?) method and leaving the update until later. There are likely to be more of these issues to resolve when updating, whereas a creating a new model based on the string contents should be straightforward.

**Comment by dladd** [comment]
One thing that has come up as I've been building this out is the question of how we treat variable units.

Currently, we are manipulating a `UnitsPtr` on a variable object with `Variable::set/getUnits()`. But I could easily see a user wanting to specify variable units using a name string, especially since the units are actually defined in the Model/Component.

**(1)** One possible solution might be to just store variable units through the name string. The existing `Variable::setUnits(UnitsPtr)` method could easily be adapted to use the name string from the argument `UnitsPtr` and we can overload a `Variable::setUnits(std::string)`. A user wanting to get the actual units of a variable would then need to call 

``` c++
libcellml::UnitsPtr units = model.getUnits(variable->getUnits());
```

**(1.a)** To be more explicit, we would probably want to rename the `Variable::set/getUnits()` methods to `set/getUnitsName()` I.e.,

``` c++
libcellml::UnitsPtr units = model.getUnits(variable->getUnitsName());
```

**(2)** As another option, we could continue to store a `UnitsPtr` on the variable and, if a user calls `Variable::setUnits(std::string)`, we can try to get the named units from the parent Model or Component. A downside is that the parent Model and Component need to exist and the named units need to exist and be added to the parent. I also see having our units objects in two places as potentially confusing.

In either case, another related issue **(X)** is that we actually don't currently have `ComponentEntity::take/get/has/replaceUnits()` methods (as we do with components). Not sure how I missed this but I can fill them out if there are no objections.  

I am for doing **(1.a)** and **(X)**.

**Comment by jonc125** [comment]
I'd be inclined to go for (2) and (X). Given that we are storing pointers, the variable would reference the same object stored on the model/component, so there wouldn't be duplication.

When serialising a model where units have been associated with a variable but not added explicitly to the parent component or model, presumably we would automatically add them to one of these; possibly the model by default (certainly if they are shared between variables in different components, which is usually the case)?

**Comment by nickerso** [comment]
I'm not sure we ever want to be adding units to a model/component serialisation automatically (in core libCellML). If a user doesn't explicitly add the units they shouldn't exist. Units are especially problematic as they could be local to a component, local to a model, or imported into the model.

The problem with having units for a variable stored as a `UnitsPtr` is that we need to define the units in order to define a variable - whereas we don't actually need a units object until such time as we need to do something with the units (validation, math, connections, etc.). I'm in favour of following some of the other recent discussions and storing properties as strings which only have meaning when an action is triggered which needs more than a straight regurgitation of the string value. This would also follow the [initial value methods](https://github.com/cellml/libcellml/blob/develop/src/api/libcellml/variable.h#L177).

So I'm in favour of (1) and (X), and I think I prefer straight out (1) to (1.a) as we are getting/setting the units property.

**Comment by agarny** [comment]
My preference is also for (1) and (X).

**Comment by jonc125** [comment]
Good points @nickerso, and I agree.

**Comment by hsorby** [comment]
On load we probably need to log inconsistencies so that the loadee can get a report on what worked.  For this to work we will need a logging class.  This would be the same class that we would use for logging messages to from the validator.

**Comment by agarny** [comment]
> On load we probably need to log inconsistencies so that the loadee can get a report on what worked.  For this to work we will need a logging class.  This would be the same class that we would use for logging messages to from the validator.

I can see this as being very useful (to report back to OpenCOR users for example), but it would clearly need some careful thinking.

For example, as an end-user, I wouldn't want to have a string telling me what is 'wrong' with this or that part of a CellML model. Instead, I would want to know the name of the CellML element that is concerned by the problem, as well as its type and the type of problem itself (and maybe other things that I haven't thought about yet).

The idea being that from this information, I should be able to refer to this problem in, say, the `Raw CellML` or `CellML Text` view in OpenCOR.

**Comment by hsorby** [comment]
At this stage we might be creating a list of errors (in string form) and the loadee would have to check if any errors were logged.  As a rough example:

```
logger::getLogger().clear();
parser.parseModel('this model');
if (logger::getLogger().msgCount()) {
  // report errors
}
```

With the view to progressing this to handing back handles to actual objects? Maybe.

**Comment by agarny** [comment]
Sure, one step at a time. I am therefore happy to go with something like your rough example to start with.

**Comment by jonc125** [comment]
Indeed. It might be worth making log 'messages' mini-objects that at present just contain a string, so it's easy to add further fields such as Alan mentions later.

**Comment by dladd** [comment]
In the thread above, we decided to go with loading new models to reduce the scope of the already sizeable chunk of work for this. But it was also acknowledged that we'd like to be able to update existing models by loading parsed attributes into them at some stage. 

Assuming all goes well after the review of #114 , am I correct in thinking we will next want to do updates via parsing before we move on to validation? If so, would others also like to see parsing/updating of other (non-model) entities or is that low-priority for now?

**Comment by hsorby** [comment]
For me parsing components and the like as well as the updating of entities is low-priority. We (with these changes) have a way to read in models the next most important thing is to be able to validate them.

**Comment by dladd** [comment]
We're almost there with finishing this up in #114 but the continuous integration has turned up one remaining snag with how we handle errors.

To recap, we've set up error objects (e.g., `VariableError`, `ModelError`) that have the shared pointers for the CellML entitites the errors are raised on. These error objects are derived classes of `EntityError`, which also has `set/getDescription()`. A user or calling environment wishing to get the object the error was raised on needs to `dynamic_cast` the `EntityError` to its derived type. For example:

``` c++
libcellml::VariableErrorPtr variableError = std::dynamic_pointer_cast<libcellml::VariableError>(parser.getError(0));
libcellml::VariablePtr variable = variableError->getVariable();
```

The current problem is that [LLVM uses its own run-time type information (RTTI)](http://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html) rather than the one built in to C++. Therefore, we can't use the above `dynamic_cast` strategy in OS X without additional work. Options I see going forward are:
- **(A)** If we want to keep the `dynamic_cast`-based strategy, we could build out the LLVM-style RTTI (examples of [developer-side implementation](http://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html#id3) and [client-side use](http://llvm.org/docs/ProgrammersManual.html#isa)). This a non-trivial bit of work to set up and requires compiler-specific checks for whether to use the c++ `dynamic_cast` or the LLVM `dyn_cast`. 
- **(B)** Alternatively, we could do away with the derived error types and just keep all flagged errors as the base `EntityError`s that we can `set/get()` the shared pointers on. Similar to the LLVM approach linked above, we can also have an `ErrorKind` enum that allows us to check the kind of error to indicate to the user/caller what object(s) to look for. However, we'd no longer need any RTTI/dynamic_casting. For example:

``` c++
if (parser.getError(0)->isKind(libcellml::ErrorKind::VariableError)) {
    libcellml::VariablePtr variable = parser.getError(0)->getVariable();
}
```

I am in favour of approach **B** as I personally find `set/get/is()` calls more user-friendly than `dynamic_cast`ing. I also think it is better to avoid compiler-specific code or work-arounds if we can for maintenance reasons.

**Comment by hsorby** [comment]
**B** for me.

**Comment by dladd** [comment]
One more thing (and hopefully the last!) that has come up when testing against the real models from cellml1to2 is that our current methods on ComponentEntities do not search through encapsulation.  For example, if `model` has the following component encapsulation:

``` c++
  <encapsulation>
    <component_ref component="main">
      <component_ref component="actual_sin"/>
    </component_ref>
  </encapsulation>
```

a call to `model.containsComponent("actual_sin")` would currently return `false`. We need to be able to get around this when parsing connections so I've implemented recursive `get/containsEncapsulatedComponent` methods to get things going for now in my branch, for example:

``` c++
ComponentPtr ComponentEntity::getEncapsulatedComponent(const std::string &name)
{
    ComponentPtr foundComponent = nullptr;
    if (containsComponent(name)) {
        foundComponent = getComponent(name);
    } else {
        for (size_t i = 0; i < componentCount(); ++i) {
            foundComponent = getComponent(i)->getEncapsulatedComponent(name);
            if (foundComponent) {
                break;
            }
        }
    }
    return foundComponent;
}
```

But the bigger question is what a user should expect the behaviour to be when they call `get/containsComponent()`. Would we **(1)** prefer the default behaviour to search through encapsulated components when we pass a name string?  Or do we **(2)** prefer to keep these methods specific to the ComponentEntity they are called from?

I can see situations where both would be useful so I'd be for **(3)** adding another set of `get/take/replace/etcEncapsulatedComponent` (as above) alongside the current `get/take/replace/etcComponent` set. I assume methods based on a passed `size_t` index will only be local, i.e. method (2).

**Comment by agarny** [comment]
Good catch about encapsulated components! Otherwise, yes, I would also favour **(3)**. However, rather than having `get/take/replace/etcEncapsulatedComponent`, why not simply have an extra parameter for `get/take/replace/etcComponent` to specify whether we want to include encapsulated components. By default, we would want to include them, hence that parameter would be defaulted to `true`.

**Comment by dladd** [comment]
Closed via #114
